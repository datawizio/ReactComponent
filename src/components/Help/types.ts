import * as React from "react";

export interface IHelpMenu {
  onTutorialLinkClick?: () => void;
  onHelperClick?: () => void;
  onVisibleChange?: (visible: boolean) => void;
  tourMenu?: React.ReactElement;
  visible?: boolean;
  tutorialDisabled?: boolean;
  helperDisabled?: boolean;
  btnDisabled?: boolean;
}