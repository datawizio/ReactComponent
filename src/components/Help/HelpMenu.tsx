import * as React from "react";
import { useContext, useMemo } from "react";
import { Menu, Dropdown, Button } from "antd";
import { QuestionCircleOutlined } from "@ant-design/icons";
import ConfigContext from "../ConfigProvider/context";
import { IHelpMenu } from "./types";

const HelpMenu: React.FC<IHelpMenu> = ({
  onTutorialLinkClick,
  onHelperClick,
  onVisibleChange,
  tourMenu,
  tutorialDisabled,
  helperDisabled,
  btnDisabled,
  visible
}) => {
  const { translate } = useContext(ConfigContext);

  const menu = useMemo(() => {
    return (
      <Menu theme="light" className="help-menu-dropdown">
        {onTutorialLinkClick && (
          <Menu.Item
            key="1"
            disabled={tutorialDisabled}
            onClick={onTutorialLinkClick}
          >
            {translate("READ_TUTORIAL")}
          </Menu.Item>
        )}
        {onHelperClick && (
          <Menu.Item key="2" disabled={helperDisabled} onClick={onHelperClick}>
            {translate("BES_HELPER")}
          </Menu.Item>
        )}
        {tourMenu}
      </Menu>
    );
  }, [
    onTutorialLinkClick,
    tutorialDisabled,
    translate,
    onHelperClick,
    helperDisabled,
    tourMenu
  ]);

  const visibleProps = typeof visible === "boolean" ? { visible } : {};

  return (
    <>
      {(onTutorialLinkClick || onHelperClick || tourMenu) && (
        <Dropdown
          {...visibleProps}
          onVisibleChange={onVisibleChange}
          overlay={menu}
          trigger={["click"]}
          placement="bottomRight"
        >
          <Button
            type="link"
            className="help-icon teaching-btn"
            onClick={e => e.preventDefault()}
            icon={<QuestionCircleOutlined />}
            disabled={btnDisabled}
          ></Button>
        </Dropdown>
      )}
    </>
  );
};

export default HelpMenu;
