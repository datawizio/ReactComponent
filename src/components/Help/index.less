@import "../../styles/variables";
@import "./rtl";

button.help-icon {
  color: @help-icon-color;
  margin-right: 16px;

  .anticon-question-circle {
    font-size: 24px;
  }

  &:hover,
  &:focus {
    color: @help-icon-color-hover;
  }
}

.help-menu-dropdown .ant-dropdown-menu-item {
  padding-top: 2px;
  padding-bottom: 2px;
}

.subjects-dropdown {
  position: fixed !important;
}
