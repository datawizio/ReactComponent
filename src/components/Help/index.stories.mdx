import { Meta, Story, Preview } from "@storybook/addon-docs/blocks";
import Help from "./index";

<Meta title="Components/Help" component={Help} />

# Help

Компонент `Help`

## Overview

<Preview>
  <Story name="Basic">
    <Help
      onTutorialClick={() => {}}
      onServiceUpdateClick={() => {}}
      onHelperClick={() => {}}
      onVisibleChange={() => {}}
      tourMenu=""
      visible={false}
    />
  </Story>
</Preview>
