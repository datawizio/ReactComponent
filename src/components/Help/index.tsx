import * as React from "react";
import HelpMenu from "./components/Menu";
import "./index.less";

export interface IHelp {
  onTutorialLinkClick?: () => void;
  onServiceUpdateClick?: () => void;
  onHelperClick?: () => void;
  onVisibleChange?: (visible: boolean) => void;
  tourMenu?: React.ReactElement;
  visible?: boolean;
}

const Help: React.FC<IHelp> = ({
  onTutorialLinkClick,
  onServiceUpdateClick,
  onHelperClick,
  onVisibleChange,
  tourMenu,
  visible
}) => {
  return (
    <>
      <HelpMenu
        onTutorialClick={onTutorialLinkClick}
        onHelperClick={onHelperClick}
        onServiceUpdateClick={onServiceUpdateClick}
        onVisibleChange={onVisibleChange}
        tourMenu={tourMenu}
        visible={visible}
      />
      <div className="divider"></div>
    </>
  );
};

export default Help;
