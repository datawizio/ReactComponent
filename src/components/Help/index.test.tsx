import "jsdom-global/register";
import React from "react";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import Help from "./index";

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  }))
});

const waitForComponentToPaint = async wrapper => {
  await act(async () => {
    await new Promise(resolve => setTimeout(resolve));
    wrapper.update();
  });
};

const mockProps = {
  onTutorialLinkClick: jest.fn(),
  onSupportLinkClick: jest.fn(),
  onServiceUpdateClick: jest.fn(),
  onSupportModalSubmit: jest.fn(),
  onHelperClick: jest.fn(),
  uploadFileURL: "url"
};

const setUp = (props?) => mount(<Help {...props} />);

describe("Help component", () => {
  let component;
  beforeEach(() => {
    component = setUp(mockProps);
  });

  it("Help rendered correctly", () => {
    expect(component).toMatchSnapshot();
  });

  it("Help rendered correctly", () => {
    const menu = component.find("HelpMenu").first();
    menu.simulate("click");
    const helper = component
      .findWhere(n => {
        return n.text() === "BES_HELPER";
      })
      .first();
    helper.simulate("click");
    waitForComponentToPaint(component);
    expect(component.prop("onHelperClick")).toBeCalledTimes(1);
  });
});
