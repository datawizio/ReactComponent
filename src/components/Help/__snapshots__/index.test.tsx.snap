// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Help component Help rendered correctly 1`] = `
<Help
  onHelperClick={[MockFunction]}
  onServiceUpdateClick={[MockFunction]}
  onSupportLinkClick={[MockFunction]}
  onSupportModalSubmit={[MockFunction]}
  onTutorialLinkClick={[MockFunction]}
  uploadFileURL="url"
>
  <HelpMenu
    onHelperClick={[MockFunction]}
    onServiceUpdateClick={[MockFunction]}
    onTutorialClick={[MockFunction]}
  >
    <Dropdown
      mouseEnterDelay={0.15}
      mouseLeaveDelay={0.1}
      overlay={
        <Menu
          className="help-menu-dropdown"
          theme="light"
        >
          <MenuItem
            onClick={[MockFunction]}
          >
            READ_TUTORIAL
          </MenuItem>
          <MenuItem
            onClick={[MockFunction]}
          >
            SERVICE_UPDATE
          </MenuItem>
          <MenuItem
            onClick={[MockFunction]}
          >
            BES_HELPER
          </MenuItem>
        </Menu>
      }
      placement="bottomRight"
      trigger={
        Array [
          "click",
        ]
      }
    >
      <ForwardRef(Dropdown)
        mouseEnterDelay={0.15}
        mouseLeaveDelay={0.1}
        overlay={[Function]}
        overlayClassName=""
        placement="bottomRight"
        prefixCls="ant-dropdown"
        transitionName="slide-up"
        trigger={
          Array [
            "click",
          ]
        }
      >
        <Trigger
          action={
            Array [
              "click",
            ]
          }
          afterPopupVisibleChange={[Function]}
          autoDestroy={false}
          blurDelay={0.15}
          builtinPlacements={
            Object {
              "bottomCenter": Object {
                "offset": Array [
                  0,
                  4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tc",
                  "bc",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "bottomLeft": Object {
                "offset": Array [
                  0,
                  4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tl",
                  "bl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "bottomRight": Object {
                "offset": Array [
                  0,
                  4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tr",
                  "br",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "topCenter": Object {
                "offset": Array [
                  0,
                  -4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "bc",
                  "tc",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "topLeft": Object {
                "offset": Array [
                  0,
                  -4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "bl",
                  "tl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "topRight": Object {
                "offset": Array [
                  0,
                  -4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "br",
                  "tr",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
            }
          }
          defaultPopupVisible={false}
          destroyPopupOnHide={false}
          focusDelay={0}
          getDocument={[Function]}
          getPopupClassNameFromAlign={[Function]}
          hideAction={Array []}
          mask={false}
          maskClosable={true}
          mouseEnterDelay={0.15}
          mouseLeaveDelay={0.1}
          onPopupAlign={[Function]}
          onPopupVisibleChange={[Function]}
          overlay={[Function]}
          popup={[Function]}
          popupAlign={Object {}}
          popupClassName=""
          popupPlacement="bottomRight"
          popupStyle={Object {}}
          popupTransitionName="slide-up"
          prefixCls="ant-dropdown"
          showAction={Array []}
          stretch="minWidth"
        >
          <Button
            className="ant-dropdown-trigger help-icon teaching-btn"
            icon={<ForwardRef(QuestionCircleOutlined) />}
            key="trigger"
            onClick={[Function]}
            onMouseDown={[Function]}
            onTouchStart={[Function]}
            type="link"
          >
            <button
              className="ant-btn ant-btn-link ant-dropdown-trigger help-icon teaching-btn"
              onClick={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
              type="button"
            >
              <ForwardRef(QuestionCircleOutlined)>
                <AntdIcon
                  icon={
                    Object {
                      "icon": Object {
                        "attrs": Object {
                          "focusable": "false",
                          "viewBox": "64 64 896 896",
                        },
                        "children": Array [
                          Object {
                            "attrs": Object {
                              "d": "M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z",
                            },
                            "tag": "path",
                          },
                          Object {
                            "attrs": Object {
                              "d": "M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0130.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1080 0 40 40 0 10-80 0z",
                            },
                            "tag": "path",
                          },
                        ],
                        "tag": "svg",
                      },
                      "name": "question-circle",
                      "theme": "outlined",
                    }
                  }
                >
                  <span
                    aria-label="question-circle"
                    className="anticon anticon-question-circle"
                    role="img"
                  >
                    <IconReact
                      icon={
                        Object {
                          "icon": Object {
                            "attrs": Object {
                              "focusable": "false",
                              "viewBox": "64 64 896 896",
                            },
                            "children": Array [
                              Object {
                                "attrs": Object {
                                  "d": "M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z",
                                },
                                "tag": "path",
                              },
                              Object {
                                "attrs": Object {
                                  "d": "M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0130.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1080 0 40 40 0 10-80 0z",
                                },
                                "tag": "path",
                              },
                            ],
                            "tag": "svg",
                          },
                          "name": "question-circle",
                          "theme": "outlined",
                        }
                      }
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="question-circle"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        key="svg-question-circle"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                          key="svg-question-circle-svg-0"
                        />
                        <path
                          d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0130.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1080 0 40 40 0 10-80 0z"
                          key="svg-question-circle-svg-1"
                        />
                      </svg>
                    </IconReact>
                  </span>
                </AntdIcon>
              </ForwardRef(QuestionCircleOutlined)>
              <span
                key=".0"
              >
                TEACHING
              </span>
            </button>
          </Button>
        </Trigger>
      </ForwardRef(Dropdown)>
    </Dropdown>
  </HelpMenu>
  <div
    className="divider"
  />
</Help>
`;
